crates/sim2h/src/demo_im_state.rs

```diff
@@ -0,0 +1,440 @@
+        con_incr: u64,
+        agent_id: AgentId,
+        aspects: im::HashSet<AspectHash>,
+    ) {
+        match self.connections.entry(agent_id.clone()) {
+            im::hashmap::Entry::Occupied(entry) => {
+                if entry.get().con_incr >= con_incr {
+                    return;
+                }
+            }
+            _ => return,
+        }
+        for aspect in aspects {
+            self.holding
+                .entry(aspect)
+                .or_default()
+                .insert(agent_id.clone());
+        }
+    }
+
+    /// return a mapping of all entry_hash/aspect_hashes
+    /// that each agent is missing (note how it returns references :+1:)
+    pub fn get_agents_missing_aspects(
+        &self,
+    ) -> im::HashMap<&AgentId, im::HashMap<&EntryHash, im::HashSet<&AspectHash>>> {
+        let mut out: im::HashMap<&AgentId, im::HashMap<&EntryHash, im::HashSet<&AspectHash>>> =
+            im::HashMap::new();
+        for (entry_hash, entry) in self.all_aspects.iter() {
+            for aspect in entry.aspects.iter() {
+                for (agent_id, _) in self.connections.iter() {
+                    if let Some(set) = self.holding.get(aspect) {
+                        if set.contains(agent_id) {
+                            continue;
+                        }
+                    }
+                    out.entry(agent_id)
+                        .or_default()
+                        .entry(entry_hash)
+                        .or_default()
+                        .insert(aspect);
+                }
+            }
+        }
+        out
+    }
+}
+
+/// wrapper around a Store clone making it look read only,
+/// without this one might be tempted to mutate something in a temporary clone
+#[derive(Clone)]
+pub struct StoreRef(Store);
+
+impl std::fmt::Debug for StoreRef {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Debug::fmt(&self.0, f)
+    }
+}
+
+impl std::ops::Deref for StoreRef {
+    type Target = Store;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::convert::AsRef<Store> for StoreRef {
+    fn as_ref(&self) -> &Store {
+        &self.0
+    }
+}
+
+impl std::borrow::Borrow<Store> for StoreRef {
+    fn borrow(&self) -> &Store {
+        &self.0
+    }
+}
+
+/// give us a cheaply clone-able async handle to the real store
+pub struct StoreHandle {
+    #[allow(dead_code)]
+    // this is just used for ref-counting
+    inner: Arc<tokio::sync::Mutex<Store>>,
+    send_mut: tokio::sync::mpsc::UnboundedSender<StoreProto>,
+    con_incr: Arc<AtomicU64>,
+}
+
+impl StoreHandle {
+    fn new(
+        inner: Arc<tokio::sync::Mutex<Store>>,
+        send_mut: tokio::sync::mpsc::UnboundedSender<StoreProto>,
+        con_incr: Arc<AtomicU64>,
+    ) -> Self {
+        Self {
+            inner,
+            send_mut,
+            con_incr,
+        }
+    }
+
+    pub async fn get_clone(&self) -> StoreRef {
```

Comment:
Is this the "non blocking" version of get_clone, then?

Reply:
yeah, exactly - in the wild you should only ever have `StoreHandle`s which let you push mutation requests across the channel and `StoreRef`s which are the immutable clones of the state...

Would be nice to have a more ergonomic single struct that could serve both purposes, but I haven't yet managed to accomplish that.